### PR TITLE
Real time output for tool deploy scripts

### LINKED ISSUE #320: Command output is buffered until finishing the whole execution
When running a deploy section (eg typical BitOps entry point), the output is buffered before finishing the command execution.

This means if Ansible or Terraform step runs `10mins`, the user will need to wait `10mins` until the subprocess closes to see the first output. Then a wall of text will output to them, making it harder to process or use in the CI/CD systems live.

In BitOps v1 the results are streamed line-by-line as we read them.
Example: https://youtu.be/TVX_87r-ySA?t=423

The same experience should be part of BitOps v2.

### PR DESCRIPTION
**Pull request by mickmcgrath13**
# Description
Uses `subprocess.Popen` instead of `subprocess.run` and uses `subprocess.PIPE` and `process.poll()` to output the subprocess execution line-by-line

Fixes #320 

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

# How Has This Been Tested?

With a local ops repo with helm charts.  Ran the `2.1.0` docker container locally and mounted the ops repo and the BitOps code to the proper locations within the container.

**Logs**
..were output in real time

**Test Configuration**:
* BitOps image tag used: 2.1.0
* Hardware: MacOS
* Tool: helm, terraform

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: scripts/plugins/deploy_plugins.py [hunk 1]
```diff
@@ -270,28 +270,60 @@ def Deploy_Plugins():
...
+
+                        # if we have a return code, exit the while loop
+                        if process.poll() is not None:
+                            break
+
+                        # if output_stdout:
+                        #     # TODO: parse output for secrets
+                        #     # TODO: specify plugin and output tight output (no extra newlines)
+                        #     # logger.info(output_stdout)
+                        #     print(output_stdout)
```
Comment (mickmcgrath13) [inline scripts/plugins/deploy_plugins.py:306]: this is probably not necessary, but I left it in because something about the implementation seems off (see previous comment)

File: scripts/plugins/deploy_plugins.py [hunk 2]
```diff
@@ -300,11 +332,11 @@ def Deploy_Plugins():
                             deployment=deployment
                         )
                     )
-                    logger.debug(result.stdout)
-                    logger.error(result.stderr)
```
Comment (mickmcgrath13) [inline scripts/plugins/deploy_plugins.py:304]: don't need these anymore because it's being output real time

File: scripts/plugins/deploy_plugins.py [hunk 3]
```diff
@@ -270,28 +270,60 @@ def Deploy_Plugins():
...
+                    #       instead, it causes the script to wait until everything is complete before outputting anything
+                    #       tried: stderr=err, stderr=subprocess.PIPE, stderr=subprocess.STDOUT
+                    #       for stderr=err, err is a file: with open(error_file,"wb") as err: 
+                    # stderr=err,
+                    # stderr=subprocess.STDOUT,
+
+                    # TODO: oddly, if you set only stderr=subprocess.PIPE and NOT stdout, 
+                    #       the process.poll() below being focused on stderr does seem to output both stdout and stderr in real time
+                    stderr=subprocess.PIPE,
+                    universal_newlines=True
```
Comment (mickmcgrath13) [inline scripts/plugins/deploy_plugins.py:291]: This was strange to me..  If we do both `stderr` and `stdout` like: 
```
stdout=subprocess.PIPE,
stderr=subprocess.PIPE
```
..the below polling doesn't seem to work.
even if stderr is one of:
- `err` (i.e. a file opened with `with open(error_file, "wb") as err:
- `subprocess.PIPE`
- `subprocess.STDOUT`

..then the weird thing is that if we do just `stdout=subprocess.PIPE,`, we do not get errors, but if we do `stderr=subprocess.PIPE`, we get both stdout and stderr real time output 🤷

**Comment by mickmcgrath13** [comment]
@armab @PhillypHenning I could use assistance figuring out what this lint error means:
https://github.com/bitovi/bitops/actions/runs/3323834010/jobs/5494675369#step:6:78

**Comment by armab** [comment]
`black` as a code formatter was added in https://github.com/bitovi/bitops/pull/297 with some documentation here: https://bitovi.github.io/bitops/development/development/#python-style-guide

@mickmcgrath13 It hints at the code `diff` [here](https://github.com/bitovi/bitops/actions/runs/3323834010/jobs/5494675369#step:6:23).

You can reformat the code with black locally: 
```
pip3 install black
black scripts/
```

Or can add it to VSCode so it reformats .py files on save automatically:
https://dev.to/adamlombard/how-to-use-the-black-python-code-formatter-in-vscode-3lo0

In other words, `black` for python is like `go fmt` for Golang.

File: scripts/plugins/deploy_plugins.py (hunk 3 above)
Comment (armab) [reply scripts/plugins/deploy_plugins.py:291]: Guessing it's just combining two streams into one if we don't need to process these streams separately.

I see where it comes from. Based on https://docs.python.org/3/library/subprocess.html#using-the-subprocess-module

> If capture_output is true, stdout and stderr will be captured. When used, the internal [Popen](https://docs.python.org/3/library/subprocess.html#subprocess.Popen) object is automatically created with stdout=PIPE and stderr=PIPE. The stdout and stderr arguments may not be supplied at the same time as capture_output. If you wish to capture and combine both streams into one, use stdout=PIPE and stderr=STDOUT instead of capture_output.

Can you try the following?
```py
process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
for combined_output in process.stdout:
    print(combined_output)
```
It worked for me in quick console tests.
That would be at least more intuitive and follow the official docs.
If works, we can reference that documentation reference then.

File: scripts/plugins/deploy_plugins.py [hunk 4]
```diff
@@ -270,28 +270,79 @@ def Deploy_Plugins():
...
+                    #         plugin_deploy_language,
+                    #         plugin_deploy_script_path,
+                    #         stack_action,
+                    #     ],
+                    #     stdout=subprocess.PIPE,
+                    #     stderr=subprocess.STDOUT
+                    # )
+
+                    # for combined_output in process.stdout:
+                    #     print(combined_output)
```
Comment (mickmcgrath13) [inline scripts/plugins/deploy_plugins.py:289]: @armab I tried this. see comment

File: scripts/plugins/deploy_plugins.py (hunk 3 above)
Comment (mickmcgrath13) [reply scripts/plugins/deploy_plugins.py]: @armab https://github.com/bitovi/bitops/pull/337#pullrequestreview-1158295846

Comment (mickmcgrath13) [reply scripts/plugins/deploy_plugins.py]: also, the quote from the link above is for `subprocess.run` not `subprocess.Popen`

File: scripts/plugins/deploy_plugins.py (hunk 4 above)
Comment (armab) [reply scripts/plugins/deploy_plugins.py]: Thanks for trying it.

https://user-images.githubusercontent.com/1533818/198304826-491502b7-7673-46b3-99e5-ca38d64c8613.mov

I will check what's missing end to end when applying this to our script.

Comment (mickmcgrath13) [reply scripts/plugins/deploy_plugins.py]: this makes me think that it might be a docker or OS ([base docker image](https://github.com/bitovi/bitops/blob/main/Dockerfile#L2)) problem 🤔

Comment (armab) [reply scripts/plugins/deploy_plugins.py]: Tried editing the original Docker container inline with that version of `deploy_plugins.py` and the live output worked, outputting terraform results line-by-line as we get it.

https://user-images.githubusercontent.com/1533818/198373559-f8db6c81-fa8b-4c13-988f-f98361d26143.mov

Make sure there's no cache involved. Worst case, we can build the temporary testing Docker images later to try it again.

**Comment by mickmcgrath13** [comment]
we tried this in the same container I'm using:
```
import subprocess
command = ["ping", "-c", "3", "1.1"]
process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.STDOUT, universal_newlines=True)
for combined_output in process.stdout:
    print(combined_output)
```

Did this by exec'ing into the container, running `python3` interactive terminal, and it worked as expected.

**Comment by mickmcgrath13** [comment]
adding `-e PYTHONUNBUFFERED=1 \` to the bitops container call seemed to help

**Comment by armab** [comment]
Yep, looks like another ENV var to hardcode in the `Dockerfile` as an addition to
https://github.com/bitovi/bitops/blob/e6f705de90408db72a373e47bc4c9db2fa9f07fb/Dockerfile#L3

**Review by PhillypHenning — APPROVED** [review]
